src/main/java/org/dependencytrack/persistence/VulnerabilityQueryManager.java

```diff
@@ -462,22 +453,22 @@ public long getVulnerabilityCount(Project project, boolean includeSuppressed) {
      * @return a List of Vulnerability objects
      */
     public List<Vulnerability> getVulnerabilities(Project project, boolean includeSuppressed) {
-        final List<Vulnerability> vulnerabilities = new ArrayList<>();
-        final List<Component> components = getAllComponents(project);
-        for (final Component component: components) {
-            final Collection<Vulnerability> componentVulns = pm.detachCopyAll(
-                    getAllVulnerabilities(component, includeSuppressed)
-            );
-            Map<String, Epss> matchedEpssList = getEpssForCveIds(
-                    componentVulns.stream().map(vuln -> vuln.getVulnId()).distinct().toList());
-            for (final Vulnerability componentVuln: componentVulns) {
-                componentVuln.setComponents(Collections.singletonList(pm.detachCopy(component)));
-                componentVuln.setAliases(new ArrayList<>(pm.detachCopyAll(getVulnerabilityAliases(componentVuln))));
-                componentVuln.setEpss(matchedEpssList.get(componentVuln.getVulnId()));
+        List<Vulnerability> projectVulnerabilities;
+        var vulnerableComponents = new HashMap<Long, Component>();
+        try (final Handle jdbiHandle = openJdbiHandle((this.request))) {
+            final var dao = jdbiHandle.attach(VulnerabilityDao.class);
+            projectVulnerabilities = dao.getVulnerabilitiesByProject(project.getId(), includeSuppressed);
+            dao.getVulnerableComponents(project.getId(), projectVulnerabilities.stream().map(Vulnerability::getId).toList())
+                    .stream().forEach(vc -> vulnerableComponents.put(vc.getId(), vc));
+        }
+        vulnerableComponents.values().forEach(vc -> vc.setProject(project));
+        for (var projectVulnerability : projectVulnerabilities) {
+            if (projectVulnerability.getComponents() != null) {
+                projectVulnerability.setComponents(
+                        projectVulnerability.getComponents().stream().map(c -> vulnerableComponents.get(c.getId())).toList());
```

Comment:
This will assign components vulnerable to **any** of the vulnerabilities, not only this particular `projectVulnerability`.

Since `dao.getVulnerableComponents` fetches components for all vulnerabilities in bulk, you're loosing context of which vulnerability a returned component is for.

You'll need a marker of sorts to tell you which component was returned for which vulnerability. One option is to additionally select `"VULNERABILITY"."ID"` in the query, and then collecting the results into a `Map<Long, List<Component>> componentsByVulnId` structure.

The above would then allow you to call this in the `for` loop:

```java
projectVulnerability.setComponents(componentsByVulnId.get(projectVulnerability.getId());
```

Reply:
List of components are fetched for each vulnerability in SQL query of `getVulnerabilitiesByProject` -> `ARRAY_AGG(DISTINCT("COMPONENT"."ID")) as "componentIdsArray"`.
Then based on this array associated with each vulnerability, components are set for those component ids `projectVulnerability.getComponents().stream().map(c -> vulnerableComponents.get(c.getId())).toList());`

Reply:
Thanks Niklas!